File: universal_printer_flutter/ios/universal_printer_flutter/Sources/universal_printer_flutter/ColumnLayout.swift
import Foundation

// Port of the Android `ColumnLayout` (com.universalprinter.util) — weighted, word-wrapped columns to
// fixed-pitch monospace lines. Kept byte-for-byte equivalent so the same row lays out identically.
enum ColumnLayout {

    static func format(_ cells: [ReceiptColumn], _ paper: PaperWidth) -> [String] {
        if cells.isEmpty { return [] }
        let clamped = Array(cells.prefix(paper.maxColumns))
        let widths = columnWidths(clamped, total: paper.charsPerLine)
        let wrapped = clamped.enumerated().map { wrap($0.element.text, width: widths[$0.offset]) }
        let rows = wrapped.map { $0.count }.max() ?? 0
        return (0..<rows).map { r in
            var line = ""
            for i in clamped.indices {
                let cell = r < wrapped[i].count ? wrapped[i][r] : ""
                line += pad(cell, width: widths[i], align: clamped[i].align)
            }
            return line
        }
    }

    static func columnWidths(_ cells: [ReceiptColumn], total: Int) -> [Int] {
        let weights = cells.map { max(1, $0.weight) }
        let sum = weights.reduce(0, +)
        var widths = weights.map { Int(Int64(total) * Int64($0) / Int64(sum)) }
        var used = widths.reduce(0, +)
        var i = 0
        while used < total && !widths.isEmpty {
            widths[i % widths.count] += 1; used += 1; i += 1
        }
        return widths
    }

    static func wrap(_ text: String, width: Int) -> [String] {
        if width <= 0 { return [""] }
        var out: [String] = []
        for rawLine in text.components(separatedBy: "\n") {
            let words = rawLine.split(whereSeparator: { $0 == " " || $0 == "\t" }).map(String.init)
            if words.isEmpty { out.append(""); continue }
            var cur = ""
            for word in words {
                var w = word
                while w.count > width {
                    if !cur.isEmpty { out.append(cur); cur = "" }
                    out.append(String(w.prefix(width)))
                    w = String(w.dropFirst(width))
                }
                if w.isEmpty { continue }
                if cur.isEmpty {
                    cur = w
                } else if cur.count + 1 + w.count <= width {
                    cur += " " + w
                } else {
                    out.append(cur); cur = w
                }
            }
            if !cur.isEmpty { out.append(cur) }
        }
        return out.isEmpty ? [""] : out
    }

    private static func pad(_ s: String, width: Int, align: TextAlign) -> String {
        let t = s.count > width ? String(s.prefix(width)) : s
        let total = width - t.count
        if total <= 0 { return t }
        switch align {
        case .left: return t + String(repeating: " ", count: total)
        case .right: return String(repeating: " ", count: total) + t
        case .center:
            let left = total / 2
            return String(repeating: " ", count: left) + t + String(repeating: " ", count: total - left)
        }
    }
}

File: universal-printer/src/main/java/com/universalprinter/util/ColumnLayout.kt
package com.universalprinter.util

import com.universalprinter.model.Align
import com.universalprinter.model.Column
import com.universalprinter.model.PaperWidth

/**
 * Turns a weighted, word-wrapped column row into monospace physical lines — rendered identically on
 * every backend (fixed-pitch Font A assumption). Cells are clamped to [PaperWidth.maxColumns], the
 * line's [PaperWidth.charsPerLine] is split by weight, each cell word-wraps to its width, and the
 * row spans as many lines as the tallest cell (shorter cells padded).
 */
internal object ColumnLayout {

    /** The composed physical lines for a column row. */
    fun format(cells: List<Column>, paper: PaperWidth): List<String> {
        if (cells.isEmpty()) return emptyList()
        val clamped = cells.take(paper.maxColumns)
        val widths = columnWidths(clamped, paper.charsPerLine)
        val wrapped = clamped.mapIndexed { i, c -> wrap(c.text, widths[i]) }
        val rows = wrapped.maxOf { it.size }
        return (0 until rows).map { r ->
            buildString {
                for (i in clamped.indices) append(pad(wrapped[i].getOrElse(r) { "" }, widths[i], clamped[i].align))
            }
        }
    }

    /** Split [total] chars across cells by weight, distributing any remainder left-to-right. */
    fun columnWidths(cells: List<Column>, total: Int): IntArray {
        val weights = cells.map { it.weight.coerceAtLeast(1) }
        val sum = weights.sum()
        val widths = IntArray(cells.size) { (total.toLong() * weights[it] / sum).toInt() }
        var used = widths.sum()
        var i = 0
        while (used < total && widths.isNotEmpty()) {
            widths[i % widths.size]++; used++; i++
        }
        return widths
    }

    /** Word-wrap [text] to [width]; breaks on whitespace, hard-breaks words longer than [width]. */
    fun wrap(text: String, width: Int): List<String> {
        if (width <= 0) return listOf("")
        val out = ArrayList<String>()
        for (rawLine in text.split("\n")) {
            val words = rawLine.split(Regex("\\s+")).filter { it.isNotEmpty() }
            if (words.isEmpty()) { out.add(""); continue }
            var cur = StringBuilder()
            for (word in words) {
                var w = word
                while (w.length > width) {
                    if (cur.isNotEmpty()) { out.add(cur.toString()); cur = StringBuilder() }
                    out.add(w.substring(0, width)); w = w.substring(width)
                }
                if (w.isEmpty()) continue
                when {
                    cur.isEmpty() -> cur.append(w)
                    cur.length + 1 + w.length <= width -> cur.append(' ').append(w)
                    else -> { out.add(cur.toString()); cur = StringBuilder(w) }
                }
            }
            if (cur.isNotEmpty()) out.add(cur.toString())
        }
        return out.ifEmpty { listOf("") }
    }

    private fun pad(s: String, width: Int, align: Align): String {
        val t = if (s.length > width) s.substring(0, width) else s
        return when (align) {
            Align.LEFT -> t.padEnd(width)
            Align.RIGHT -> t.padStart(width)
            Align.CENTER -> {
                val total = width - t.length
                val left = total / 2
                " ".repeat(left) + t + " ".repeat(total - left)
            }
        }
    }
}
